# Ejercicios - Práctica 7a: Procesos en el *shell* de Linux.

<br>
<font size="5">  
    <span style="color: red;"><b>ATENCIÓN:</b><br>Se recomienda realizar los ejercicios propuestos de este boletín en un terminal independiente, ya que la mayoría de ellos <b>no pueden llevarse a cabo en Jupyter</b> debido a que es necesario interactuar con el programa lanzado mediante el teclado durante su ejecución. 
    </span>
    <br><br>
    Utiliza el <i>notebook</i> únicamente para responder a las preguntas o introducir el comando requerido en las celdas de código.
</font>  


### Control de procesos en primer y segundo plano

- Lanza el programa `gnome-text-editor` desde un terminal, y luego pulsa `Ctrl-C` para enviar al proceso una señal de terminación. ¿Puedes terminar su ejecución de esta forma? ¿Por qué? ¿A qué proceso le llega la señal `SIGINT` enviada al pulsar `Ctrl-C`?



In [ ]:
RESPUESTA: No, no se puede terminar la ejecución de gnome-text-editor utilizando Ctrl-C desde el terminal. La señal SIGINT enviada al pulsar Ctrl-C llega al proceso en primer plano asociado al terminal

Vuelve a lanzar el mismo programa, esta vez en segundo plano, y tras hacerlo, pulsa de nuevo `Ctrl-C`. ¿Puedes terminar su ejecución así? ¿Por qué? ¿A qué proceso le llega esta vez la señal `SIGINT`?



In [ ]:
# Comando para lanzar gnome-text-editor en segundo plano
gnome-text-editor &


- ¿Qué comando utilizarías para poder terminar el comando lanzado anteriormente sabiendo su PID?

In [ ]:
# Comando para terminar la ejecución del proceso lanzado a partir del nombre del programa
kill <PID>


PREGUNTA: Una vez terminada la ejecución del comando anterior, sin teclear ningún comando, presiona repetidamente `Ctrl-C`. ¿Qué ocurre cada vez que lo haces? ¿Es posible terminar la ejecución del shell de esa forma?


RESPUESTA: Ctrl-C envía una señal al proceso en ejecución en primer plano. No se puede terminar la ejecución de gnome-text-editor usando Ctrl-C si este está ejecutándose en un entorno gráfico de GNOME o similar.



- Lanza los programas `okteta` y `gnome-text-editor` desde el terminal, ambos en primer plano, como una secuencia de comandos consecutivos, en la misma línea de la siguiente celda:

In [ ]:
# Lanzar okteta y después gnome-text-editor, ambos en primer plano
okteta; gnome-text-editor

PREGUNTA: ¿Qué ha de ocurrir para que comience a ejecutarse el editor `gnome-text-editor`? ¿Cuándo se completa la ejecución de la celda del notebook?

RESPUESTA: hasta que no he cerrado okteta no se ha ejecutado gnome-text-editor



- Vuelve a lanzar los dos programas en una misma línea de comandos, pero esta vez en segundo plano.

In [ ]:
# Lanzar okteta y después gnome-text-editor, ambos en segundo plano
okteta & gnome-text-editor &


- Escribe el comando adecuado para pasar a primer plano el último proceso que se lanzó (el editor `gnome-text-editor`).

RESPUESTA: fg



PREGUNTA: Desde el terminal, termina el proceso que acabas de pasar a primer plano. ¿Qué teclas o comando has utilizado?

RESPUESTA: ctrl-c



PREGUNTA: Repite el comando `fg` para hacer lo propio con el comando anterior (`okteta`), pero esta vez, en lugar de terminarlo, páralo. ¿Qué teclas o comando has utilizado?

RESPUESTA:  ctrl-z



PREGUNTA: Trata de interactuar con el proceso `okteta`, ¿por que no responde al ratón ni al teclado?

RESPUESTA: Cuando un proceso está suspendido, se detiene completamente y no puede responder a ninguna interacción, ya sea mediante el ratón o el teclado. 



PREGUNTA: Vuelve a despertar el proceso `okteta` pero esta vez en segundo plano. ¿Qué comando has de utilizar?

RESPUESTA: bg %1



- Usa la orden adecuada para listar los procesos lanzados desde el terminal donde estás ejecutando los comandos anteriores.

In [1]:
# Comando para monitorizar los procesos activos lanzados en este terminal
jobs

```
    PID TTY          TIME CMD
  19422 pts/5    00:00:00 bash
  27254 pts/5    00:00:01 okteta
  28051 pts/5    00:00:00 ps
```

- Finalmente, termina el proceso `okteta` desde el terminal esta vez usando el PID (identificador del proceso) que nos indicó el terminal al lanzar el proceso inicialmente.

In [2]:
# Comando para terminar el proceso lanzado en segundo plano
 kill 34655

bash: kill: (34655) - No existe el proceso


: 1

PREGUNTA: Lanza de nuevo los programas `okteta` y `gnome-text-editor`, en dicho orden, pero esta vez únicamente con `okteta` en segundo plano. ¿Qué ocurre si ahora pulsas `Ctrl-C` en el terminal?

RESPUESTA: se cierra el gnome-text-editor



- Usando el comando adecuado, averigua el PID del proceso que está ejecutando `okteta` a partir de su nombre.

In [ ]:
# Comando para averiguar el PID del un proceso por el nombre del comando ejecutado
pgrep okteta

- Termina el programa `okteta` sin necesidad de introducir su PID

In [3]:
pkill okteta


: 1

PREGUNTA: Descarga el programa ejecutable `uninterruptible`, el cual tienes disponible a través del AV, y cámbiale los permisos para poder ejecutarlo. Ejecútalo en primer plano y prueba a terminar su ejecución con `CTRL-C`. ¿Puedes? Explica qué semejanzas crees que tiene este programa con el programa *shell* en cuanto al tratamiento de la señal `SIGINT`.

RESPUESTA:  no consigo instalar ni ejecutar el probrama uninterruptible



PREGUNTA: Detén la ejecución del programa lanzado anteriormente, sin terminarlo. Después lista los procesos lanzados en el terminal, ¿cuánto tiempo ha usado la CPU el programa `uninterruptible` antes de ser detenido? Si vuelves a listar los procesos, ¿ha variado el tiempo de CPU?

RESPUESTA:



PREGUNTA: Reanuda en segundo plano la ejecución del programa detenido anteriormente, y usa el comando adecuado para observar periódicamente el tiempo de CPU que lleva consumido. ¿Qué porcentaje de la CPU usa cuando está en ejecución? ¿En qué medida crece el tiempo de uso de la CPU con respecto al tiempo "del reloj de pared" (medido por una persona)?

RESPUESTA:



PREGUNTA: Termina el programa `uninterruptible`, ya sea por su nombre o PID, enviando la señal `SIGTERM`. ¿Puedes terminar el proceso? Indica el comando que has utilizado.

RESPUESTA:



PREGUNTA: A la vista de la salida del programa `uninterruptible` terminado en el paso anterior, ¿crees que el programa termina súbitamente al recibir la señal, o lo hace por sus propios medios?

RESPUESTA:



PREGUNTA: Descarga el programa `stubborn` del AV y cámbiale los permisos para poder ejecutarlo. Después, ejecútalo en primer plano y prueba a terminar su ejecución con `CTRL-C`. ¿Qué ocurre? De nuevo, observa los procesos en ejecución con el comando adecuado, ¿cuál es su uso de CPU?

RESPUESTA:



PREGUNTA: Detén la ejecución del programa lanzado anteriormente, sin terminarlo, y luego reanuda en segundo plano su ejecución. Después, trata de terminar el programa `stubborn`, ya sea por su nombre o PID, enviando la señal `SIGTERM` al igual que has hecho anteriormente. ¿Puedes terminar el proceso de esta forma?

RESPUESTA:



PREGUNTA: A la vista de lo anterior, ¿de qué forma podemos terminar el proceso `stubborn`? Indica el comando.

RESPUESTA:

La única forma de acabar con el proceso es "matarlo" con `pkill -9 stubborn`, lo cual hace que se le envíe la señal `SIGKILL`, que ningún proceso puede ignorar. Al hacerlo, vemos este mensaje: 
```
[1]+  Terminado (killed)      ./stubborn
```


### Listado y monitorización de procesos

- Usa el comando adecuado para mostrar **todos** los procesos que se están ejecutando actualmente en el sistema.

In [4]:
ps aux


USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root           1  0.0  0.0 167672 12740 ?        Ss   nov14   4:31 /sbin/init
root           2  0.0  0.0      0     0 ?        S    nov14   0:00 [kthreadd]
root           3  0.0  0.0      0     0 ?        S    nov14   0:00 [pool_workque
root           4  0.0  0.0      0     0 ?        I<   nov14   0:00 [kworker/R-rc
root           5  0.0  0.0      0     0 ?        I<   nov14   0:00 [kworker/R-rc
root           6  0.0  0.0      0     0 ?        I<   nov14   0:00 [kworker/R-sl
root           7  0.0  0.0      0     0 ?        I<   nov14   0:00 [kworker/R-ne
root          10  0.0  0.0      0     0 ?        I<   nov14   0:00 [kworker/0:0H
root          12  0.0  0.0      0     0 ?        I<   nov14   0:00 [kworker/R-mm
root          13  0.0  0.0      0     0 ?        I    nov14   0:00 [rcu_tasks_ru
root          14  0.0  0.0      0     0 ?        I    nov14   0:00 [rcu_tasks_tr
root          15  0.0  0.0      0     0 

- Localiza tres procesos cuyo propietario sea nuestro usuario, y otros tres cuyo propietario sea el superusuario (root). Pega la línea completa mostrada para cada proceso.

RESPUESTA:

root           1  0.0  0.0 167672 12740 ?        Ss   nov14   4:31 /sbin/init

root           2  0.0  0.0      0     0 ?        S    nov14   0:00 [kthreadd]

root           3  0.0  0.0      0     0 ?        S    nov14   0:00 [pool_workque

jupyter+  680694  0.1  0.3 672824 54092 ?        Ssl  11:26   0:01 /opt/tljh/use

jupyter+  680706  0.0  0.0  11588  5632 pts/14   Ss+  11:26   0:00 /usr/bin/bash

jupyter+  681020  0.1  0.3 672824 53580 ?        Ssl  11:31   0:00 /opt/tljh/use




- Localiza el proceso que ha consumido más tiempo de CPU. Pega la línea completa mostrada.

RESPUESTA:  root         472  0.1  1.0 980720 175772 ?       Ssl  nov14  52:41 /opt/tljh/hub




- Localiza también el último que se lanzó, mirando las horas de inicio. Pega la línea completa mostrada.

RESPUESTA:  
jupyter+  681403  0.0  0.0  12776  3584 pts/23   R+   11:38   0:00 ps aux




- Lanza desde la interfaz de usuario el navegador firefox, y navega hasta la página `www.google.es`, mientras observas mediante `top` la evolución dinámica del consumo de memoria y CPU de los procesos involucrados.  Puedes dividir la pantalla entre ambas ventanas, terminal y navegador, para observar la salida de top mientras navegas. Luego, prueba a navegar a la página `www.youtube.com`. ¿En cuál de las dos páginas se hace más uso de la CPU?

RESPUESTA: En YouTube se hace más uso de la CPU.



### Redireccionamiento de la entrada/salida de un proceso

In [5]:
# Lista todos los ficheros en `/usr/bin/` cuyo nombre empiece por 'l'
# y tenga exactamente dos caracteres
ls /usr/bin/l? 


/usr/bin/ld  /usr/bin/ln  /usr/bin/ls


In [6]:
# Repite el comando anterior, redireccionando la salida a un fichero llamado progs-l.txt
ls /usr/bin/l? > progs-l.txt


In [7]:
# Muestra el contenido del fichero progs-l.txt
cat progs-l.txt

/usr/bin/ld
/usr/bin/ln
/usr/bin/ls


In [8]:
# Añade al final de dicho fichero un mensaje que diga "FIN DEL LISTADO", usando `echo` y la redirección
echo "FIN DEL LISTADO" >> progs-l.txt


In [9]:
# Muestra de nuevo el contenido del fichero
cat progs-l.txt

/usr/bin/ld
/usr/bin/ln
/usr/bin/ls
FIN DEL LISTADO


In [10]:
# Utilizando el comando ls con la opción adecuada y la redirección de la salida estándar/de error,
# averigua sobre qué subdirectorios de /etc no tienes permisos de acceso. PISTA: usa /dev/null
# para descartar la salida que no te interese.

ls -l /etc | grep "^d" 2>/dev/null



drwxr-xr-x 2 root root       4096 sep  3 02:56 alternatives
drwxr-xr-x 3 root root       4096 sep  3 02:04 apache2
drwxr-xr-x 3 root root       4096 oct  8 06:56 apparmor
drwxr-xr-x 8 root root       4096 oct  8 06:56 apparmor.d
drwxr-xr-x 3 root root       4096 ago 29 17:24 apport
drwxr-xr-x 8 root root       4096 ago 29 17:29 apt
drwxr-xr-x 2 root root       4096 ago 29 17:24 bash_completion.d
drwxr-xr-x 2 root root       4096 abr  7  2022 binfmt.d
drwxr-xr-x 2 root root       4096 ago 29 17:24 byobu
drwxr-xr-x 3 root root       4096 ago 29 17:22 ca-certificates
drwxr-xr-x 4 root root       4096 sep  3 02:03 chrony
drwxr-xr-x 5 root root       4096 sep  5 21:59 cloud
drwxr-xr-x 2 root root       4096 ago 29 17:24 console-setup
drwxr-xr-x 2 root root       4096 ago 29 17:23 cron.d
drwxr-xr-x 2 root root       4096 nov 14 10:11 cron.daily
drwxr-xr-x 2 root root       4096 ago 29 17:22 cron.hourly
drwxr-xr-x 2 root root       4096 ago 29 17:22 cron.monthly
drwxr-xr-x 2 root root       4

In [11]:
# Crea dos directorios, llamados dir1 y dir2
mkdir dir1 dir2

In [20]:
# Copia el fichero stubborn a dir2
cp stubborn dir2/
#no encuentro el fichero stubborn


cp: no se puede efectuar `stat' sobre 'stubborn': No existe el archivo o el directorio


: 1

In [21]:
# Listar el contenido de dir1 y dir2 en un mismo comando
ls dir1 dir2

dir1:

dir2:


In [22]:
# Cambia los permisos a dir2 para que nadie tenga acceso
chmod 000 dir2

In [23]:
# Listar el contenido de dir1 y dir2 en un mismo comando
ls dir1 dir2

dir1:
ls: no se puede abrir el directorio 'dir2': Permiso denegado


: 2

In [24]:
#a salida Utilizando la redirección de l estándar/de error, lista
# los directorios dir1 y dir2, de forma que los mensajes de error
# vayan al fichero "errores_ls" y el resto al fichero "salida_ls"
ls dir1 dir2 > salida_ls 2> errores_ls


: 2

In [25]:
# Muestra el contenido del fichero errores_ls
cat errores_ls

ls: no se puede abrir el directorio 'dir2': Permiso denegado


In [26]:
# Muestra el contenido del fichero salida_ls
cat salida_ls

dir1:


### Variables de entorno. Concepto de PATH

- La variable de entorno `LANG` define, junto con otras, la configuración regional y el idioma del shell. Empieza mostrando su valor actual con `echo`.

In [29]:
echo $LANG


es_ES.UTF-8


 - Ejecuta el comando `ls` pasándole una ruta inexistente para generar un mensaje de error (coméntalo tras generarlo).

In [30]:
ls /ruta/inexistente
#ls: no se puede acceder a '/ruta/inexistente': No existe el archivo o el directorio

ls: no se puede acceder a '/ruta/inexistente': No existe el archivo o el directorio


: 2

-  Ahora, cambia el valor de la variable `LANG` por "xyz" (da igual el valor)  y seguidamente muestra su valor con `echo`.

In [31]:
export LANG=xyz
echo $LANG


xyz


 - Vuelve a ejecutar el comando `ls` pasándole una ruta inexistente para generar un mensaje de error (coméntalo # de nuevo).

In [33]:
ls /ruta/inexistente
#ls: cannot access '/ruta/inexistente': No such file or directory

ls: cannot access '/ruta/inexistente': No such file or directory


: 2

PREGUNTA:  ¿En qué se diferencia el resultado de ejecutar el comando con un valor u otro de `$LANG`? ¿Cuál es el idioma por defecto del shell cuando $LANG no tiene un valor?

RESPUESTA: Si $LANG tiene un valor como "xyz" el sistema puede no aplicar correctamente la localización, y algunos comportamientos predeterminados (como el idioma de los mensajes) se mantendrán en inglés o un valor predeterminado.



- Vuelve a establecer el valor de `LANG` a su valor original ("es_ES.UTF-8") y seguidamente muestra su valor con `echo` (en una misma línea).

In [34]:
export LANG=es_ES.UTF-8
echo $LANG


es_ES.UTF-8


In [46]:
# Crea un subdirectorio llamado bin en tu directorio de inicio de usuario (si no existe y

mkdir bin

In [47]:
# Copia al directorio creado anteriormente el fichero /usr/bin/ls, con el nombre myls
cp /usr/bin/ls bin/myls


In [50]:
# Sin cambiarte de directorio, ejecuta mediante una ruta absoluta el programa `myls` con la opción '-l':
/home/jupyter-7MwOMGMO2n/bin/myls -l


total 182220
drwxr-xr-x  2 jupyter-7MwOMGMO2n jupyter-7MwOMGMO2n      4096 dic  9 12:17 bin
drwxr-xr-x 13 jupyter-7MwOMGMO2n jupyter-7MwOMGMO2n      4096 nov 28 12:24 fc-alumno
drwxrwxr-x  4 jupyter-7MwOMGMO2n jupyter-7MwOMGMO2n      4096 sep  7  2023 ficheros_ejemplo
-rw-r--r--  1 jupyter-7MwOMGMO2n jupyter-7MwOMGMO2n 186547355 sep 26 10:28 ficheros_ejemplo.zip
drwxr-xr-x  3 jupyter-7MwOMGMO2n jupyter-7MwOMGMO2n      4096 oct 14 15:56 geografia-murciana
drwxr-xr-x  6 jupyter-7MwOMGMO2n jupyter-7MwOMGMO2n      4096 sep 23 12:39 myfirstrepo
drwxr-xr-x  3 jupyter-7MwOMGMO2n jupyter-7MwOMGMO2n      4096 oct 14 13:22 p2-ejercicio
drwxr-xr-x  3 jupyter-7MwOMGMO2n jupyter-7MwOMGMO2n      4096 nov  7 17:11 p4b-ejercicios
drwxr-xr-x  3 jupyter-7MwOMGMO2n jupyter-7MwOMGMO2n      4096 oct 24 12:53 pruebas-bash
drwxr-xr-x  4 jupyter-7MwOMGMO2n jupyter-7MwOMGMO2n      4096 oct 14 15:55 test-repo
drwxr-xr-x  3 jupyter-7MwOMGMO2n jupyter-7MwOMGMO2n      4096 oct 14 16:09 tmp


In [51]:
# Modifica el valor de `PATH` para añadir el subdirectorio "bin" creado anteriormente:
export PATH=$PATH:/home/jupyter-7MwOMGMO2n/bin/myls 


Finalmente, prueba a ejecutar `myls` sin necesidad de indicar su ruta exacta.

In [53]:
myls -l


Orden «myls» no encontrada. Quizá quiso decir:
  la orden «tyls» del paquete deb «terminology (1.12.1-1)»
  la orden «mmls» del paquete deb «sleuthkit (4.11.1+dfsg-1)»
  la orden «pyls» del paquete deb «python3-pyls (0.36.2-3)»
Pruebe con: apt install <nombre del paquete deb>


: 127

### Variables del shell actual y variables de entorno

- Muestra el valor de la variable de entorno `USER`

In [54]:
echo $USER


jupyter-7MwOMGMO2n


In [55]:
# Define una nueva variable llamada "PRUEBA" con valor "hola"
export PRUEBA="hola"


In [56]:
# Muestra el valor de la variable PRUEBA en el shell actual
echo $PRUEBA


hola


In [57]:
# Ejecuta esta celda para crear un shell script llamado "variables.sh"
echo '#!/bin/bash' > variables.sh
echo 'echo "**Inicio del script**"' >> variables.sh
echo 'echo "La variable PRUEBA vale $PRUEBA"' >> variables.sh
echo 'MYVAR="hello world!"' >> variables.sh
echo 'echo "La variable MYVAR vale $MYVAR"' >> variables.sh
echo 'echo "La variable USER vale $USER"' >> variables.sh
echo 'echo "La variable ANOTHERVAR vale $ANOTHERVAR "' >> variables.sh
echo 'echo "**Fin del script**"' >> variables.sh

In [58]:
# Muestra el contenido del script "variables.sh"
cat variables.sh


#!/bin/bash
echo "**Inicio del script**"
echo "La variable PRUEBA vale $PRUEBA"
MYVAR="hello world!"
echo "La variable MYVAR vale $MYVAR"
echo "La variable USER vale $USER"
echo "La variable ANOTHERVAR vale $ANOTHERVAR "
echo "**Fin del script**"


In [59]:
# Establece permisos de ejecución para el fichero creado
chmod +x variables.sh



In [66]:
# Ejecuta el script "variables.sh"
./variables.sh

**Inicio del script**
La variable PRUEBA vale hola
La variable MYVAR vale hello world!
La variable USER vale jupyter-7MwOMGMO2n
La variable ANOTHERVAR vale  
**Fin del script**


PREGUNTA: ¿Se produce algún error en el shell cuando tratas de mostrar el valor de una variable a la que no has asignado valor previamente (ANOTHERVAR)?

RESPUESTA: si, no se muestra su valor



PREGUNTA: ¿Por qué no se imprime el valor de `PRUEBA`?

RESPUESTA: en mi caso si se imprime



In [67]:
# Imprime el valor de la variable MYVAR en el shell actual
echo $MYVAR


PREGUNTA: ¿Por qué no aparece "hello world" como valor de "MYVAR" en el shell actual, tras ejecutar `variables.sh`?

RESPUESTA: Las variables definidas dentro de un script por defecto solo existen en el entorno del script y no afectan al entorno del shell actual.



PREGUNTA: ¿Qué valor tiene la variable "USER" en el proceso creado para ejecutar el script `variables.sh`?

RESPUESTA: jupyter-7MwOMGMO2n




**FIN DEL CUADERNO** (*no borres esta celda*)

@@@@ practica7a-procesos-ejercicios.ipynb @@@@